pandas/io/stata.py

```diff
@@ -1979,11 +1979,16 @@ def _open_file_binary_write(
         compression_typ, compression_args = get_compression_method(compression)
         compression_typ = infer_compression(fname, compression_typ)
         compression = dict(compression_args, method=compression_typ)
-        path_or_buf, _, compression, _ = get_filepath_or_buffer(
+        ioargs = get_filepath_or_buffer(
             fname, mode="wb", compression=compression, storage_options=storage_options,
         )
-        f, _ = get_handle(path_or_buf, "wb", compression=compression, is_text=False)
-        return f, True, compression
+        f, _ = get_handle(
+            ioargs.filepath_or_buffer,
+            "wb",
+            compression=ioargs.compression,
+            is_text=False,
+        )
+        return f, True, ioargs.compression
```

Comment:
might need a slight refactoring but can you use IOargs here?

Reply:
it returns similar information but in this case `f` is always `BinaryIO` (according to the current type annotations). If we were to use IOargs, we would give it a broader type.

Reply:
I think the following options would make sense for this function:

1. inline it (it is used only one time and it is a private function)
2. simplify it (the first if-block is imho not necessary, can be handled by the elif-block)
3. simplification from (2) and move the function to io/common and use it more than once: the pattern to convert the compression string/dict to a dict, calling `get_filepath_or_buffer`, and then calling `get_handle` is present in many `to_*` (and `read_*`) functions.

I'm tempted to go for the second option in this PR and leave the third option for a future PR.

Reply:
I will not touch that for now: I don't understand the difference between `self._output_file` and `self._file` in `StataWriter`. Option (2) would affect these two variables. Having had only a brief look at `StataWriter`, it seems that this class distinguishes between writing to a compressed file and writing to a buffer (`get_handle` should take care of that, after that you should be able to treat them the same?).

Reply:
yeah this can all addressed later. this writer might need some restructuing.